### PR TITLE
Add typing to samplers

### PR DESCRIPTION
**Pull request by JoelPasvolsky**
Needs a careful review please

### CODE REVIEW

File: tests/requirements.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 coverage[toml]
 codecov
 
+mypy==0.942
```
Comment (arcondello) [inline tests/requirements.txt:4]: We don't actually test with this, so we can leave it out for now.

File: dwave/samplers/greedy/sampler.py [hunk 2]
```diff
@@ -114,32 +119,34 @@ def __init__(self):
             'large_sparse_opt_values': (True, False),
         }
 
-    def sample(self, bqm, num_reads=None, initial_states=None,
-               initial_states_generator="random", seed=None,
-               large_sparse_opt=False, **kwargs):
+    def sample(self, bqm: dimod.BinaryQuadraticModel,
+               num_reads: Optional[int] = None,
+               initial_states: Optional[dimod.typing.SamplesLike] = None,
+               initial_states_generator: InitialStateGenerator = "random",
```
Comment (arcondello) [inline dwave/samplers/greedy/sampler.py:125]: Let's actually leave ` InitialStateGenerator` as `str` for now. Eventually we should import the type from dimod for future proofing.

File: dwave/samplers/sa/sampler.py [hunk 3]
```diff
@@ -17,6 +17,13 @@
 from numbers import Integral
 from numpy.random import randint
 from collections import defaultdict
+from typing import (List, Literal, Tuple, Optional, Union)
```
Comment (arcondello) [inline dwave/samplers/sa/sampler.py:20]:
```python
from typing import List, Tuple, Optional, Union

try:
    from typing import Literal
    BetaScheduleType = Literal['linear', 'geometric', 'custom']
except ImportError:
    BetaScheduleType = str

```

see comment above for handling initial states.

File: dwave/samplers/sa/sampler.py [hunk 4]
```diff
@@ -123,40 +130,46 @@ def __init__(self):
...
-               **kwargs):
+    def sample(self, bqm: dimod.BinaryQuadraticModel,
+               beta_range: Optional[Union[List[float], Tuple[float, float]]] = None,
+               num_reads: Optional[int] = None,
+               num_sweeps: Optional[int] = None,
+               num_sweeps_per_beta: int = 1,
+               beta_schedule_type: BetaScheduleType = "geometric",
+               seed: Optional[int] = None,
+               interrupt_function=None,
+               beta_schedule: Optional[np.ndarray] = None,
```
Comment (arcondello) [inline dwave/samplers/sa/sampler.py:141]: `Union[Sequence[float], np.ndarray]`. We could use `np.ArrayLike` but that's actually way more than we want.

File: dwave/samplers/greedy/sampler.py (hunk 2 above)
Comment (arcondello) [reply dwave/samplers/greedy/sampler.py:125]: Oh wait, you can!

```
from dimod.core.samplers.initializated import InitialStateGenerator
```

Comment (JoelPasvolsky) [reply dwave/samplers/greedy/sampler.py:125]: Oh good, I looked for this in dimod hoping we had it but saw [local implementation](https://github.com/dwavesystems/dimod/blob/6ee9013f4281ef9d7752fbaa4a35be1d42bb9ebc/dimod/reference/samplers/identity_sampler.py#L24) and figured no one had done that yet

Comment (arcondello) [reply dwave/samplers/greedy/sampler.py:125]: That should really be moved into `dimod.typing` at some point... But that's a future PR.

File: tests/requirements.txt (hunk 1 above)
Comment (JoelPasvolsky) [reply tests/requirements.txt:4]: Removed mypy from test requirements

**Comment by arcondello** [comment]
Do you also intend to do the properties/parameters?

**Comment by JoelPasvolsky** [comment]
> Do you also intend to do the properties/parameters?

I think I just need to change to ``properties: dict = None``, right? I'll update them

**Comment by arcondello** [comment]
> I think I just need to change to `properties: dict = None`, right? I'll update them

Mostly to drop it from the strings. I actually think you can skip the type definition since they are inherited from the ABC.

**Pull request merged**